clang/include/clang/Sema/Sema.h

```diff
@@ -6898,10 +6898,18 @@ class Sema final {
                                    BinaryOperatorKind Operator);
 
   //// ActOnCXXThis -  Parse 'this' pointer.
-  ExprResult ActOnCXXThis(SourceLocation loc);
+  ///
+  /// \param SkipLambdaCaptureCheck Whether to skip the 'this' check for a
+  /// lambda because 'this' is the lambda's 'this'-pointer.
+  ExprResult ActOnCXXThis(SourceLocation loc,
+                          bool SkipLambdaCaptureCheck = false);
```

Comment:
Sure, do you have any suggestions? 
From what I understood, the previous implementation always checked whether capturing `this` was valid. My patch adds support for lambdas in coroutines with custom `new`. It needs a `this'- pointer but not a captured one of the surrounding class; the lambdas own `this'- pointer. Skipping the capture check sounded appropriate.

Reply:
Maybe something like, `IsNonStaticLambda`?  But I guess this only applies in a coroutine situation, so perhaps that is imperfect too.  I might have to think it through.

Perhaps @ChuanqiXu9 can come up with something.

Reply:
Maybe `IsLambdasOwnThis`?

Reply:
I can't parse that... can you describe that name a little more?

Reply:
It would require invert the logic in the code but state that the `this'- pointer that `ActOnCXXThis` handles is the lambda's own `this` and not the one from the lambda's surrounding scope. I don't know a more CWG wording term for the lambdas `this`- pointer.

Reply:
Can we provide more rationale for skipping lambda's capture check here? For example, in what cases and for what reasons, we want to skip the check. I am still confused for that.

Reply:
Sure, have a look at this code (https://cppinsights.io/s/3212c7be)

```
class Test {
  int a;

public:
  Test(int x)
  : a{x}
  {
    int other{};
    auto l1 = [=] { return a + 2 + other; };
  }
};
```

The lambda `l1` has _two_ things that we refer to as `this`-pointer. It's very own, as every non-static member function has. Used to access `other` like `this->other` inside the lambdas body. That `this`-pointer isn't captured. Hence, no check.

Then the captured `this`-pointer from the enclosing scope of `Test`. In C++ Insights called `__this`. While being used it looks like `this->__this->a`.

The check is in place for the latter. For the use case of calling `operator new` in the case of a coroutine, we are talking about the first one, which isn't captured.

While writing about this, I haven't checked whether a `promise_type` with a constructor worked before or does so now. I can check that maybe tomorrow.

Reply:
Excellent. Good. I suggest you to update this to the inline comments.

Reply:
If there is any problem with the promise_type with a constructor, let's make it in other patches.

Reply:
How about `ThisRefersToClosureObject` ?

Reply:
Sounds good! Thanks!